### PR TITLE
Refactoring raft log module

### PR DESCRIPTION
**Pull request by LebronAl**
Testing and waiting reviewing~

### CODE REVIEW

File: cluster/src/main/java/org/apache/iotdb/cluster/log/UnCommittedEntryManager.java [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+        if (low < offset) {
+            logger.debug("unCommittedEntryManager getEntries[{},{}) out of bound : [{},{}]", low, high, offset, upper);
+            low = offset;
+        }
+        if (high > upper) {
+            logger.info("unCommittedEntryManager getEntries[{},{}) out of bound : [{},{}] , adjust parameter 'high' to {}", low, high, offset, upper, upper);
+            // don't throw a exception to support getEntries(low, Integer.MAX_VALUE).
+            high = upper;
+        }
+        return entries.subList((int) (low - offset), (int) (high - offset));
```
Comment (SilverNarcissus) [inline cluster/src/main/java/org/apache/iotdb/cluster/log/UnCommittedEntryManager.java:181]: If the caller change the subList, our manager will be impacted. You can use a new arrayList to return or add some comment at the method

Comment (LebronAl) [reply cluster/src/main/java/org/apache/iotdb/cluster/log/UnCommittedEntryManager.java:181]: Actually,In RaftLogManager's getEntries implemetation,a new list is created to deeply copy these references.So it is safe to change the list outside RaftLogManager.
Thanks for reviewing!

File: cluster/src/main/java/org/apache/iotdb/cluster/log/manage/UnCommittedEntryManager.java [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+     * method is only called after persisting newly committed entries.
+     *
+     * @param index request entry's index
+     * @param term  request entry's term
+     */
+    public void stableTo(long index, long term) {
+        try {
+            long entryTerm = maybeTerm(index);
+            // only update the uncommitted entries if term is matched with an uncommitted entry.
+            if (entryTerm == term) {
```
[the remark this answers is not in the document]
Comment (LebronAl) [reply cluster/src/main/java/org/apache/iotdb/cluster/log/manage/UnCommittedEntryManager.java:105]: This check is not required under the current design. I will delete it~

File: cluster/src/test/java/org/apache/iotdb/cluster/log/manage/RaftLogManagerTest.java [hunk 3]
```diff
@@ -0,0 +1,787 @@
...
+public class RaftLogManagerTest {
+
+	private Set<Log> appliedLogs;
+	private LogApplier logApplier = new TestLogApplier() {
+		@Override
+		public void apply(Log log) {
+			appliedLogs.add(log);
+		}
+	};
+
```
[the remark this answers is not in the document]
Comment (LebronAl) [reply cluster/src/test/java/org/apache/iotdb/cluster/log/manage/RaftLogManagerTest.java:51]: OK~

File: cluster/src/main/java/org/apache/iotdb/cluster/log/manage/RaftLogManager.java [hunk 4]
```diff
@@ -0,0 +1,420 @@
...
+    }
+
+    /**
+     * Used by leader node or MaybeAppend to directly append to unCommittedEntryManager. Note that the
+     * caller should ensure entries[0].index > committed.
+     *
+     * @param entries appendingEntries
+     * @return the newly generated lastIndex
+     */
+    public long append(List<Log> entries) {
```
[the remark this answers is not in the document]
Comment (LebronAl) [reply cluster/src/main/java/org/apache/iotdb/cluster/log/manage/RaftLogManager.java:210]: I think that as long as batch replication is supported, any log will be sent in batches to follower, and the single-log-append interface will only be used for leader to append member changes logs as long as the client can also send a batch of logs~But for extensibility,  I have added a single-log-append  interface~

File: cluster/src/main/java/org/apache/iotdb/cluster/server/member/DataGroupMember.java [hunk 5]
```diff
@@ -918,7 +916,9 @@ public boolean closePartition(String storageGroupName, long partitionId, boolean
       log.setPreviousLogTerm(logManager.getLastLogTerm());
       log.setCurrLogIndex(logManager.getLastLogIndex() + 1);
 
-      logManager.appendLog(log);
+      logManager.append(new ArrayList<Log>() {{
+        add(log);
+      }});
```
[the remark this answers is not in the document]
Comment (LebronAl) [reply cluster/src/main/java/org/apache/iotdb/cluster/server/member/DataGroupMember.java:921]: OK~I will fix it.

File: cluster/src/main/java/org/apache/iotdb/cluster/server/member/RaftMember.java [hunk 6]
```diff
@@ -671,17 +672,22 @@ public void catchUp(Node follower, long followerLastLogIndex) {
...
         if (!allLogsValid) {
           // if the first log has been snapshot, the snapshot should also be sent to the
           // follower, otherwise some data will be missing
           snapshot = logManager.getSnapshot();
+        }else{
+          try {
+            logs = logManager.getEntries(followerLastLogIndex, Long.MAX_VALUE);
+          }catch (Exception e){
+              logger.error("Unexpected error: {}" ,e.getMessage());
+          }
```
[the remark this answers is not in the document]
Comment (LebronAl) [reply cluster/src/main/java/org/apache/iotdb/cluster/server/member/RaftMember.java:690]: Sorry,I misunderstood the code here. I thought the logs weren't needed when the snapshot was sent.I will fix this~

File: cluster/src/main/java/org/apache/iotdb/cluster/log/manage/CommittedEntryManager.java [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+     * Pack entries from low through high - 1, just like slice (entries[low:high]). dummyIndex < low
+     * <= high. Note that caller must ensure low <= high.
+     *
+     * @param low  request index low bound
+     * @param high request index upper bound
+     * @throws EntryCompactedException
+     */
+    public List<Log> getEntries(long low, long high) throws EntryCompactedException {
+        if (low > high) {
+            logger.debug("invalid getEntries: parameter: {} > {}", low, high);
```
[the remark this answers is not in the document]
Comment (LebronAl) [reply cluster/src/main/java/org/apache/iotdb/cluster/log/manage/CommittedEntryManager.java:131]: OK~I will return a empty list here~

File: cluster/src/main/java/org/apache/iotdb/cluster/log/manage/UnCommittedEntryManager.java [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+     *
+     * @param low  request index low bound
+     * @param high request index upper bound
+     */
+    public List<Log> getEntries(long low, long high) {
+        if (low > high) {
+            logger
+                .debug("invalid unCommittedEntryManager getEntries: parameter: low({}) > high({})",
+                    low, high);
+        }
```
[the remark this answers is not in the document]
Comment (LebronAl) [reply cluster/src/main/java/org/apache/iotdb/cluster/log/manage/UnCommittedEntryManager.java:171]: Fine~

**Pull request merged**